docs/design/003-wasm-support.md

```diff
@@ -0,0 +1,69 @@
+# Wasm Support in Inspektor Gadget
+
+We want to further extend the flexibility of [image-based
+gadgets](./002-containerized-gadgets.md) by allowing users to run custom code in
+user space.
+
+## Wasm Capabilities
+
+This section covers the features Wasm should provide to gadget developers.
+
+### Manipulate Data Sources
+
+The Wasm module should be able to manipulate the data coming from other sources,
+(eBPF being one of them). It'll allow the Wasm module to change the format of
+the fields, for instance, performing string manipulations that are difficult to
+implement in eBPF, to enrich events with new information it gathers from other
+places, etc. The Wasm program should also be able to create a new data source on
+its own to provide new information.
+
+This is the list of operations the Wasm module should be able to make:
+
+- Change the value of fields
+  - Implement data manipulations impossible to make in eBPF
+- Add fields
+  - Create "virtual fields" by combining existing ones. Like the "call" field on the trace mount gadget
+  - Create new fields to enrich the event with information the Wasm module has
+- Change fields attributes
+  - Hide / show the field
+  - Change the width used to print it
+  - etc.
+- Drop events
+  - Implement advanced filtering capabilities
+- Create new data sources
+
+### Handle eBPF objects
+
+There are cases when it's useful to allow the Wasm module to directly interact
+with eBPF objects.
+
+- Manipulate maps: Get, Delete, Update, etc. (many possible operations allowed on a map)
+- Manipulate the eBPF objects before loading
+  - Disable / enable programs based on kernel features.
```

Comment:
That's good, thanks for adding this.

Some related issues:
* https://github.com/alban/fsnotify-enricher/issues/4
* https://github.com/cilium/ebpf/discussions/1470

IG could offer the following functions to the WASM guest:
* pkg/kallsyms/kallsyms.go
  * KAllSymsSymbolExists(name string) bool
  * KAllSymsLookupByInstructionPointer(ip uint64) string
  * KAllSymsSpecUpdateAddresses(funcName string) error
* btf
  * BTFKernelSpecTypeByName(name string) (btfID u32) # more complicated: how to marshall btf.Type?
* Enable/Disable
  * ELFSectionDisableAutoAttach(sectionName string)
  * ELFSectionManualAttach(sectionName string, attachTo string)

The fsnotify-enricher gadget could do something like:
```go
if !KAllSymsSymbolExists("fsnotify_insert_event") {
  ELFSectionDisableAutoAttach("kprobe/fsnotify_insert_event")
  ELFSectionManualAttach("kprobe/fsnotify_insert_event", "fsnotify_add_event")
}
```

ELFSectionManualAttach could be called several times to implement something like pwru.

cc @i-Pear

Reply:
> ELFSectionDisableAutoAttach(sectionName string)

I think maybe `progName` is better, because it's the name of function whichis unique. For different kernel versions, you may have many functions with the same section name (attaching point).

Reply:
Thanks @alban for the suggestions, I'm adding them to the document. I used BPFProgram instead of ELFSection as this is the terminology used by libbpf -> https://github.com/libbpf/libbpf/blob/9b789075a993e8235aac301126e58ff7b380da5f/src/libbpf.h#L334